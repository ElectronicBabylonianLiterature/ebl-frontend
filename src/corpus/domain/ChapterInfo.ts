import { TextLine } from 'transliteration/domain/text-line'
import TranslationLine from 'transliteration/domain/translation-line'
import { Line } from 'corpus/domain/line'
import { ChapterId } from 'transliteration/domain/chapter-id'

export interface ChapterInfoLine extends Omit<Line, 'translation'> {
  translation: ReadonlyArray<TranslationLine>
}

export default interface ChapterInfo {
  readonly id: ChapterId
  readonly textName: string
  readonly siglums: Record<string, string>
  readonly matchingLines: readonly ChapterInfoLine[]
  readonly matchingColophonLines: Record<string, readonly TextLine[]>
}
