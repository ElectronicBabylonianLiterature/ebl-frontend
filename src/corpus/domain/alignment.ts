export type AlignmentToken =
  | {
      value: string
    }
  | {
      value: string
      alignment: number | null
      variant: ''
    }
  | {
      value: string
      alignment: number | null
      variant: string
      language: string
      isNormalized: boolean
    }

export type Alignment = readonly AlignmentToken[][][]
