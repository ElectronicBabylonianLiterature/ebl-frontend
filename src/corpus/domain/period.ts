export type PeriodModifier = {
  readonly name: string
  readonly displayName: string
}
export const periodModifiers: ReadonlyMap<string, PeriodModifier> = new Map([
  ['None', { name: 'None', displayName: '-' }],
  ['Early', { name: 'Early', displayName: 'Early' }],
  ['Late', { name: 'Late', displayName: 'Late' }],
])

export type Period = {
  readonly name: string
  readonly abbreviation: string
  readonly description: string
  readonly parent?: string
  readonly displayName?: string
}
export const periods: ReadonlyMap<string, Period> = new Map([
  [
    'None',
    {
      name: 'None',
      abbreviation: '',
      description: '',
      displayName: '-',
    },
  ],
  [
    'Ur III',
    {
      name: 'Ur III',
      abbreviation: 'Ur3',
      description: '(ca. 2100–2002 BCE)',
    },
  ],
  [
    'Old Assyrian',
    {
      name: 'Old Assyrian',
      abbreviation: 'OA',
      description: '(ca. 1950–1850 BCE)',
    },
  ],
  [
    'Old Babylonian',
    {
      name: 'Old Babylonian',
      abbreviation: 'OB',
      description: '(ca. 2002–1595 BCE)',
    },
  ],
  [
    'Middle Babylonian',
    {
      name: 'Middle Babylonian',
      abbreviation: 'MB',
      description: '(ca. 1595–1000 BCE)',
    },
  ],
  [
    'Middle Assyrian',
    {
      name: 'Middle Assyrian',
      abbreviation: 'MA',
      description: '(ca. 1595–1000 BCE)',
    },
  ],
  [
    'Hittite',
    {
      name: 'Hittite',
      abbreviation: 'Hit',
      description: '(ca. 1500–1100 BCE)',
    },
  ],
  [
    'Neo-Assyrian',
    {
      name: 'Neo-Assyrian',
      abbreviation: 'NA',
      description: '(ca. 1000–609 BCE)',
    },
  ],
  [
    'Neo-Babylonian',
    {
      name: 'Neo-Babylonian',
      abbreviation: 'NB',
      description: '(ca. 1000–539 BCE)',
    },
  ],
  [
    'Late Babylonian',
    {
      name: 'Late Babylonian',
      abbreviation: 'LB',
      description: '(ca. 539 BCE–ca. 100 CE)',
    },
  ],
  [
    'Persian',
    {
      name: 'Persian',
      abbreviation: 'Per',
      description: '(539–331 BCE)',
      parent: 'Late Babylonian',
    },
  ],
  [
    'Hellenistic',
    {
      name: 'Hellenistic',
      abbreviation: 'Hel',
      description: '(331–141 BCE)',
      parent: 'Late Babylonian',
    },
  ],
  [
    'Parthian',
    {
      name: 'Parthian',
      abbreviation: 'Par',
      description: '(141 BCE–ca. 100 CE)',
      parent: 'Late Babylonian',
    },
  ],
  ['Uncertain', { name: 'Uncertain', abbreviation: 'Unc', description: '' }],
])
