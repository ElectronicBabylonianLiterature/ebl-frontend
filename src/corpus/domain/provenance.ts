export const Provenances = {
  'Standard Text': { name: 'Standard Text', abbreviation: 'Std', parent: null },
  Assyria: { name: 'Assyria', abbreviation: 'Assa', parent: null },
  Aššur: { name: 'Aššur', abbreviation: 'Ašš', parent: 'Assyria' },
  Ḫuzirina: { name: 'Ḫuzirina', abbreviation: 'Huz', parent: 'Assyria' },
  Kalḫu: { name: 'Kalḫu', abbreviation: 'Kal', parent: 'Assyria' },
  Khorsabad: { name: 'Khorsabad', abbreviation: 'Kho', parent: 'Assyria' },
  Nineveh: { name: 'Nineveh', abbreviation: 'Nin', parent: 'Assyria' },
  Tarbiṣu: { name: 'Tarbiṣu', abbreviation: 'Tar', parent: 'Assyria' },
  Babylonia: { name: 'Babylonia', abbreviation: 'Baba', parent: null },
  Babylon: { name: 'Babylon', abbreviation: 'Bab', parent: 'Babylonia' },
  Borsippa: { name: 'Borsippa', abbreviation: 'Bor', parent: 'Babylonia' },
  Cutha: { name: 'Cutha', abbreviation: 'Cut', parent: 'Babylonia' },
  Dilbat: { name: 'Dilbat', abbreviation: 'Dil', parent: 'Babylonia' },
  Isin: { name: 'Isin', abbreviation: 'Isn', parent: 'Babylonia' },
  Kiš: { name: 'Kiš', abbreviation: 'Kiš', parent: 'Babylonia' },
  Larsa: { name: 'Larsa', abbreviation: 'Lar', parent: 'Babylonia' },
  Meturan: { name: 'Meturan', abbreviation: 'Met', parent: 'Babylonia' },
  Nērebtum: { name: 'Nērebtum', abbreviation: 'Nēr', parent: 'Babylonia' },
  Nippur: { name: 'Nippur', abbreviation: 'Nip', parent: 'Babylonia' },
  Sippar: { name: 'Sippar', abbreviation: 'Sip', parent: 'Babylonia' },
  Šaduppûm: { name: 'Šaduppûm', abbreviation: 'Šad', parent: 'Babylonia' },
  Ur: { name: 'Ur', abbreviation: 'Ur', parent: 'Babylonia' },
  Uruk: { name: 'Uruk', abbreviation: 'Urk', parent: 'Babylonia' },
  Periphery: { name: 'Periphery', abbreviation: '', parent: null },
  Alalakh: { name: 'Alalakh', abbreviation: 'Ala', parent: 'Periphery' },
  'Tell el-Amarna': {
    name: 'Tell el-Amarna',
    abbreviation: 'Ama',
    parent: 'Periphery',
  },
  Emar: { name: 'Emar', abbreviation: 'Emr', parent: 'Periphery' },
  Ḫattuša: { name: 'Ḫattuša', abbreviation: 'Hat', parent: 'Periphery' },
  Mari: { name: 'Mari', abbreviation: 'Mar', parent: 'Periphery' },
  Megiddo: { name: 'Megiddo', abbreviation: 'Meg', parent: 'Periphery' },
  Susa: { name: 'Susa', abbreviation: 'Sus', parent: 'Periphery' },
  Ugarit: { name: 'Ugarit', abbreviation: 'Uga', parent: 'Periphery' },
  Uncertain: { name: 'Uncertain', abbreviation: 'Unc', parent: null },
} as const
export type Provenance = typeof Provenances[keyof typeof Provenances]
export const provenances = [
  Provenances['Standard Text'],
  Provenances.Assyria,
  Provenances.Aššur,
  Provenances.Ḫuzirina,
  Provenances.Kalḫu,
  Provenances.Khorsabad,
  Provenances.Nineveh,
  Provenances.Tarbiṣu,
  Provenances.Babylonia,
  Provenances.Babylon,
  Provenances.Borsippa,
  Provenances.Cutha,
  Provenances.Dilbat,
  Provenances.Isin,
  Provenances.Kiš,
  Provenances.Larsa,
  Provenances.Meturan,
  Provenances.Nērebtum,
  Provenances.Nippur,
  Provenances.Sippar,
  Provenances.Šaduppûm,
  Provenances.Ur,
  Provenances.Uruk,
  Provenances.Periphery,
  Provenances.Alalakh,
  Provenances['Tell el-Amarna'],
  Provenances.Emar,
  Provenances.Ḫattuša,
  Provenances.Mari,
  Provenances.Megiddo,
  Provenances.Susa,
  Provenances.Ugarit,
  Provenances.Uncertain,
] as const

export function compareStandardText(
  first: Provenance,
  second: Provenance
): number {
  if (first === second) {
    return 0
  } else if (first === Provenances['Standard Text']) {
    return -1
  } else if (second === Provenances['Standard Text']) {
    return 1
  } else {
    return 0
  }
}

export function compareAssyriaAndBabylonia(
  first: Provenance,
  second: Provenance
): number {
  function isCity(provenance: Provenance): boolean {
    const nonCities: Provenance[] = [
      Provenances['Standard Text'],
      Provenances.Assyria,
      Provenances.Babylonia,
    ]
    return !nonCities.includes(provenance)
  }
  if (isCity(first) && isCity(second)) {
    return 0
  } else if (isCity(first)) {
    return 1
  } else if (isCity(second)) {
    return -1
  } else {
    return compareName(first, second)
  }
}

export function compareName(first: Provenance, second: Provenance): number {
  return first.name.localeCompare(second.name)
}
