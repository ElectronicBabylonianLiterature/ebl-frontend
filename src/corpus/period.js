import type { RecordFactory, RecordOf } from 'immutable'
import { OrderedMap, Record } from 'immutable'

type PeriodModifierProps = { name: string, displayName: string }
const createPeriodModifier: RecordFactory<PeriodModifierProps> = Record({
  name: '',
  displayName: ''
})
export type PeriodModifier = RecordOf<PeriodModifierProps>
export const periodModifiers: OrderedMap<string, PeriodModifier> = OrderedMap({
  None: createPeriodModifier({ name: 'None', displayName: '-' }),
  Early: createPeriodModifier({ name: 'Early', displayName: 'Early' }),
  Late: createPeriodModifier({ name: 'Late', displayName: 'Late' })
})

type PeriodProps = {
  name: string,
  abbreviation: string,
  description: string,
  parent: ?string
}
const createPeriod: RecordFactory<PeriodProps> = Record({
  name: '',
  abbreviation: '',
  description: '',
  parent: null
})
export type Period = RecordOf<PeriodProps>
export const periods: OrderedMap<string, Period> = OrderedMap({
  'Ur III': createPeriod({
    name: 'Ur III',
    abbreviation: 'Ur3',
    description: '(ca. 2100–2002 BCE)'
  }),
  'Old Assyrian': createPeriod({
    name: 'Old Assyrian',
    abbreviation: 'OA',
    description: '(ca. 1950–1850 BCE)'
  }),
  'Old Babylonian': createPeriod({
    name: 'Old Babylonian',
    abbreviation: 'OB',
    description: '(ca. 2002–1595 BCE)'
  }),
  'Middle Babylonian': createPeriod({
    name: 'Middle Babylonian',
    abbreviation: 'MB',
    description: '(ca. 1595–1000 BCE)'
  }),
  'Middle Assyrian': createPeriod({
    name: 'Middle Assyrian',
    abbreviation: 'MA',
    description: '(ca. 1595–1000 BCE)'
  }),
  Hittite: createPeriod({
    name: 'Hittite',
    abbreviation: 'Hit',
    description: '(ca. 1500–1100 BCE)'
  }),
  'Neo-Assyrian': createPeriod({
    name: 'Neo-Assyrian',
    abbreviation: 'NA',
    description: '(ca. 1000–609 BCE)'
  }),
  'Neo-Babylonian': createPeriod({
    name: 'Neo-Babylonian',
    abbreviation: 'NB',
    description: '(ca. 1000–539 BCE)'
  }),
  'Late Babylonian': createPeriod({
    name: 'Late Babylonian',
    abbreviation: 'LB',
    description: '(ca. 539 BCE–ca. 100 CE)'
  }),
  Persian: createPeriod({
    name: 'Persian',
    abbreviation: 'Per',
    description: '(539–331 BCE)',
    parent: 'Late Babylonian'
  }),
  Hellenistic: createPeriod({
    name: 'Hellenistic',
    abbreviation: 'Hel',
    description: '(331–141 BCE)',
    parent: 'Late Babylonian'
  }),
  Parthian: createPeriod({
    name: 'Parthian',
    abbreviation: 'Par',
    description: '(141 BCE–ca. 100 CE)',
    parent: 'Late Babylonian'
  }),
  Uncertain: createPeriod({
    name: 'Uncertain',
    abbreviation: 'Unc',
    description: ''
  })
})
