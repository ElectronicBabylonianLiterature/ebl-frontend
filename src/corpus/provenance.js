import type { RecordFactory, RecordOf } from 'immutable'
import { OrderedMap, Record } from 'immutable'

type ProvenanceProps = { name: string, abbreviation: string, parent: ?string }
const createProvenance: RecordFactory<ProvenanceProps> = Record({
  name: '',
  abbreviation: '',
  parent: null
})
export type Provenance = RecordOf<ProvenanceProps>
export const provenances: OrderedMap<string, Provenance> = OrderedMap({
  Assyria: createProvenance({ name: 'Assyria', abbreviation: 'Assa' }),
  Aššur: createProvenance({
    name: 'Aššur',
    abbreviation: 'Ašš',
    parent: 'Assyria'
  }),
  Ḫuzirina: createProvenance({
    name: 'Ḫuzirina',
    abbreviation: 'Huz',
    parent: 'Assyria'
  }),
  Kalḫu: createProvenance({
    name: 'Kalḫu',
    abbreviation: 'Kal',
    parent: 'Assyria'
  }),
  Khorsabad: createProvenance({
    name: 'Khorsabad',
    abbreviation: 'Kho',
    parent: 'Assyria'
  }),
  Nineveh: createProvenance({
    name: 'Nineveh',
    abbreviation: 'Nin',
    parent: 'Assyria'
  }),
  Tarbiṣu: createProvenance({
    name: 'Tarbiṣu',
    abbreviation: 'Tar',
    parent: 'Assyria'
  }),
  Babylonia: createProvenance({ name: 'Babylonia', abbreviation: 'Baba' }),
  Babylon: createProvenance({
    name: 'Babylon',
    abbreviation: 'Bab',
    parent: 'Babylonia'
  }),
  Borsippa: createProvenance({
    name: 'Borsippa',
    abbreviation: 'Bor',
    parent: 'Babylonia'
  }),
  Cutha: createProvenance({
    name: 'Cutha',
    abbreviation: 'Cut',
    parent: 'Babylonia'
  }),
  Isin: createProvenance({
    name: 'Isin',
    abbreviation: 'Isn',
    parent: 'Babylonia'
  }),
  Kiš: createProvenance({
    name: 'Kiš',
    abbreviation: 'Kiš',
    parent: 'Babylonia'
  }),
  Larsa: createProvenance({
    name: 'Larsa',
    abbreviation: 'Lar',
    parent: 'Babylonia'
  }),
  Meturan: createProvenance({
    name: 'Meturan',
    abbreviation: 'Met',
    parent: 'Babylonia'
  }),
  Nērebtum: createProvenance({
    name: 'Nērebtum',
    abbreviation: 'Nēr',
    parent: 'Babylonia'
  }),
  Nippur: createProvenance({
    name: 'Nippur',
    abbreviation: 'Nip',
    parent: 'Babylonia'
  }),
  Sippar: createProvenance({
    name: 'Sippar',
    abbreviation: 'Sip',
    parent: 'Babylonia'
  }),
  Šaduppûm: createProvenance({
    name: 'Šaduppûm',
    abbreviation: 'Šad',
    parent: 'Babylonia'
  }),
  Ur: createProvenance({ name: 'Ur', abbreviation: 'Ur', parent: 'Babylonia' }),
  Uruk: createProvenance({
    name: 'Uruk',
    abbreviation: 'Urk',
    parent: 'Babylonia'
  }),
  Periphery: createProvenance({ name: 'Periphery', abbreviation: '' }),
  Alalakh: createProvenance({
    name: 'Alalakh',
    abbreviation: 'Ala',
    parent: 'Periphery'
  }),
  'Tell el-Amarna': createProvenance({
    name: 'Tell el-Amarna',
    abbreviation: 'Ama',
    parent: 'Periphery'
  }),
  Emar: createProvenance({
    name: 'Emar',
    abbreviation: 'Emr',
    parent: 'Periphery'
  }),
  Ḫattuša: createProvenance({
    name: 'Ḫattuša',
    abbreviation: 'Hat',
    parent: 'Periphery'
  }),
  Mari: createProvenance({
    name: 'Mari',
    abbreviation: 'Mar',
    parent: 'Periphery'
  }),
  Megiddo: createProvenance({
    name: 'Megiddo',
    abbreviation: 'Meg',
    parent: 'Periphery'
  }),
  Susa: createProvenance({
    name: 'Susa',
    abbreviation: 'Sus',
    parent: 'Periphery'
  }),
  Ugarit: createProvenance({
    name: 'Ugarit',
    abbreviation: 'Uga',
    parent: 'Periphery'
  }),
  Uncertain: createProvenance({ name: 'Uncertain', abbreviation: 'Unc' })
})
