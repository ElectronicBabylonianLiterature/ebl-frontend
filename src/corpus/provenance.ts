export type Provenance = {
  readonly name: string
  readonly abbreviation: string
  readonly parent?: string
}
export const provenances: ReadonlyMap<string, Provenance> = new Map([
  ['Assyria', { name: 'Assyria', abbreviation: 'Assa' }],
  ['Aššur', { name: 'Aššur', abbreviation: 'Ašš', parent: 'Assyria' }],
  ['Ḫuzirina', { name: 'Ḫuzirina', abbreviation: 'Huz', parent: 'Assyria' }],
  ['Kalḫu', { name: 'Kalḫu', abbreviation: 'Kal', parent: 'Assyria' }],
  ['Khorsabad', { name: 'Khorsabad', abbreviation: 'Kho', parent: 'Assyria' }],
  ['Nineveh', { name: 'Nineveh', abbreviation: 'Nin', parent: 'Assyria' }],
  ['Tarbiṣu', { name: 'Tarbiṣu', abbreviation: 'Tar', parent: 'Assyria' }],
  ['Babylonia', { name: 'Babylonia', abbreviation: 'Baba' }],
  ['Babylon', { name: 'Babylon', abbreviation: 'Bab', parent: 'Babylonia' }],
  ['Borsippa', { name: 'Borsippa', abbreviation: 'Bor', parent: 'Babylonia' }],
  ['Cutha', { name: 'Cutha', abbreviation: 'Cut', parent: 'Babylonia' }],
  ['Isin', { name: 'Isin', abbreviation: 'Isn', parent: 'Babylonia' }],
  ['Kiš', { name: 'Kiš', abbreviation: 'Kiš', parent: 'Babylonia' }],
  ['Larsa', { name: 'Larsa', abbreviation: 'Lar', parent: 'Babylonia' }],
  ['Meturan', { name: 'Meturan', abbreviation: 'Met', parent: 'Babylonia' }],
  ['Nērebtum', { name: 'Nērebtum', abbreviation: 'Nēr', parent: 'Babylonia' }],
  ['Nippur', { name: 'Nippur', abbreviation: 'Nip', parent: 'Babylonia' }],
  ['Sippar', { name: 'Sippar', abbreviation: 'Sip', parent: 'Babylonia' }],
  ['Šaduppûm', { name: 'Šaduppûm', abbreviation: 'Šad', parent: 'Babylonia' }],
  ['Ur', { name: 'Ur', abbreviation: 'Ur', parent: 'Babylonia' }],
  ['Uruk', { name: 'Uruk', abbreviation: 'Urk', parent: 'Babylonia' }],
  ['Periphery', { name: 'Periphery', abbreviation: '' }],
  ['Alalakh', { name: 'Alalakh', abbreviation: 'Ala', parent: 'Periphery' }],
  [
    'Tell el-Amarna',
    { name: 'Tell el-Amarna', abbreviation: 'Ama', parent: 'Periphery' },
  ],
  ['Emar', { name: 'Emar', abbreviation: 'Emr', parent: 'Periphery' }],
  ['Ḫattuša', { name: 'Ḫattuša', abbreviation: 'Hat', parent: 'Periphery' }],
  ['Mari', { name: 'Mari', abbreviation: 'Mar', parent: 'Periphery' }],
  ['Megiddo', { name: 'Megiddo', abbreviation: 'Meg', parent: 'Periphery' }],
  ['Susa', { name: 'Susa', abbreviation: 'Sus', parent: 'Periphery' }],
  ['Ugarit', { name: 'Ugarit', abbreviation: 'Uga', parent: 'Periphery' }],
  ['Uncertain', { name: 'Uncertain', abbreviation: 'Unc' }],
])
