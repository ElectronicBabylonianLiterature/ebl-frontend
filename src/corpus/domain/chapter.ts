import _ from 'lodash'
import { immerable } from 'immer'
import flow from 'lodash/fp/flow'
import flatMap from 'lodash/fp/flatMap'
import map from 'lodash/fp/map'
import Cite from 'citation-js'
import removeMd from 'remove-markdown'
import { Parser } from 'html-to-react'
import { LineNumber, OldLineNumber } from 'transliteration/domain/line-number'
import { MarkupPart } from 'transliteration/domain/markup'
import { ChapterAlignment } from './alignment'
import { Line, ManuscriptLine } from './line'
import { Manuscript } from './manuscript'
import { TextId, textIdToDoiString } from 'transliteration/domain/text-id'
import TranslationLine from 'transliteration/domain/translation-line'
import { ChapterId, defaultName } from 'transliteration/domain/chapter-id'
import { NoteLine } from 'transliteration/domain/note-line'
import { ParallelLine } from 'transliteration/domain/parallel-line'
import { Token } from 'transliteration/domain/token'
import { ManuscriptLineDisplay } from 'corpus/domain/line-details'
import { stageToAbbreviation } from 'corpus/domain/period'

export class Chapter {
  readonly [immerable] = true

  constructor(
    readonly textId: TextId,
    readonly textHasDoi: boolean,
    readonly classification: string,
    readonly stage: string,
    readonly version: string,
    readonly name: string,
    readonly order: number,
    readonly manuscripts: ReadonlyArray<Manuscript>,
    readonly uncertainFragments: ReadonlyArray<string>,
    readonly lines: ReadonlyArray<Line>
  ) {}

  get id(): ChapterId {
    return {
      textId: this.textId,
      stage: this.stage,
      name: this.name,
    }
  }

  get alignment(): ChapterAlignment {
    return new ChapterAlignment(
      this.lines.map((line) =>
        line.variants.map((variant) => variant.alignment)
      )
    )
  }

  getSiglum(manuscriptLine: ManuscriptLine): string {
    const manuscript = this.manuscripts.find(
      (candidate) => candidate.id === manuscriptLine.manuscriptId
    )
    if (manuscript) {
      return manuscript.siglum
    } else {
      return `<unknown ID: ${manuscriptLine.manuscriptId}>`
    }
  }
}

export interface LineVariantDisplay {
  readonly reconstruction: ReadonlyArray<Token>
  readonly note: NoteLine | null
  readonly manuscripts: ReadonlyArray<ManuscriptLineDisplay>
  readonly parallelLines: ReadonlyArray<ParallelLine>
  readonly intertext: ReadonlyArray<MarkupPart>
}

export interface LineDisplay {
  readonly number: LineNumber
  readonly oldLineNumbers: ReadonlyArray<OldLineNumber>
  readonly isSecondLineOfParallelism: boolean
  readonly isBeginningOfSection: boolean
  readonly translation: ReadonlyArray<TranslationLine>
  readonly variants: ReadonlyArray<LineVariantDisplay>
}

export interface Author {
  readonly name: string
  readonly prefix: string
  readonly role: 'EDITOR' | 'REVISION'
  readonly orcidNumber: string
}

export interface Translator {
  readonly name: string
  readonly prefix: string
  readonly orcidNumber: string
  readonly language: string
}

export interface Record {
  readonly authors: ReadonlyArray<Author>
  readonly translators: ReadonlyArray<Translator>
  readonly publicationDate: string
}

export class ChapterDisplay {
  readonly [immerable] = true

  constructor(
    readonly id: ChapterId,
    readonly textHasDoi: boolean,
    readonly textName: string,
    readonly isSingleStage: boolean,
    readonly title: ReadonlyArray<MarkupPart>,
    readonly lines: ReadonlyArray<LineDisplay>,
    readonly record: Record,
    readonly atf: string
  ) {}

  get languages(): Set<string> {
    return flow(
      flatMap<LineDisplay, TranslationLine>((line) => line.translation),
      map((translation) => translation.language),
      (languages) => new Set(languages)
    )(this.lines)
  }

  get isPublished(): boolean {
    return (
      !_.isEmpty(this.record.publicationDate) && !_.isEmpty(this.record.authors)
    )
  }

  get uniqueIdentifier(): string {
    return this.idParts.join(' ')
  }

  get fullName(): string {
    const showName = this.id.name !== defaultName || this.isSingleStage
    return [
      removeMd(this.textName),
      'Chapter',
      !this.isSingleStage ? this.id.stage : '',
      showName ? this.id.name : '',
    ]
      .filter(_.negate(_.isEmpty))
      .join(' ')
  }

  get url(): string {
    const idPartsAbbreviation = [
      ...this.idParts.slice(0, 3),
      stageToAbbreviation(this.idParts[3]),
      this.idParts[4],
    ]
    return `https://www.ebl.lmu.de/corpus/${idPartsAbbreviation
      .map(encodeURIComponent)
      .join('/')}`
  }

  get doi(): string {
    return this.textHasDoi ? textIdToDoiString(this.id.textId) : ''
  }

  get parsedCitation(): JSX.Element {
    return new Parser().parse(
      this.citation.format('bibliography', {
        format: 'html',
        template: 'citation-apa',
      })
    )
  }

  get citation(): Cite {
    const issued = new Date(this.record.publicationDate)
    const now = new Date()
    return new Cite({
      id: this.uniqueIdentifier,
      type: 'article-journal',
      author: this.record.authors.map((author) => ({
        family: author.name,
        given: author.prefix,
      })),
      accessed: {
        'date-parts': [[now.getFullYear(), now.getMonth() + 1, now.getDate()]],
      },
      issued: {
        'date-parts': [
          [issued.getFullYear(), issued.getMonth() + 1, issued.getDate()],
        ],
      },
      title: this.fullName,
      'container-title': 'electronic Babylonian Literature',
      URL: this.url,
      ...(this.doi && { DOI: this.doi }),
    })
  }

  private get idParts(): readonly [string, number, number, string, string] {
    return [
      this.id.textId.genre,
      this.id.textId.category,
      this.id.textId.index,
      this.id.stage,
      this.id.name,
    ]
  }

  getTranslatorsOf(language: string): Translator[] {
    return this.record.translators.filter(
      (translator) => translator.language === language
    )
  }
}
