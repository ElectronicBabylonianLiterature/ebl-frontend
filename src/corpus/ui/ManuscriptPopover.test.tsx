import React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import ManuscriptPopOver from './ManuscriptPopover'
import { manuscriptLineDisplayFactory } from 'test-support/line-details-fixtures'
import userEvent from '@testing-library/user-event'
import { MemoryRouter } from 'react-router-dom'
import { Provenances } from 'corpus/domain/provenance'
import { Periods } from 'common/period'
import Chance from 'chance'

const chance = new Chance('manuscript-popover-test')

const manuscript = manuscriptLineDisplayFactory.build(
  {},
  {
    associations: {
      provenance: Provenances.Babylon,
      period: Periods['Late Babylonian'],
    },
    transient: { chance: chance },
  }
)
const oldSiglum = manuscript.oldSigla[0]

beforeEach(() =>
  render(
    <MemoryRouter>
      <ManuscriptPopOver manuscript={manuscript} />
    </MemoryRouter>
  )
)

test('Open manuscript line popover', async () => {
  const siglumText = screen.getByText(manuscript.siglum)
  expect(siglumText).toBeVisible()

  userEvent.click(siglumText)
  await waitFor(() => expect(screen.getByRole('tooltip')).toBeVisible())
})

test('Show manuscript line details', async () => {
  userEvent.click(screen.getByText(manuscript.siglum))
  await waitFor(() => expect(screen.getByRole('tooltip')).toBeVisible())

  const heading = screen.getByRole('heading', { level: 3 })
  expect(heading).toHaveTextContent(oldSiglum.siglum)
  expect(heading).toMatchSnapshot()

  const number = manuscript.joins[0][0].museumNumber
  expect(screen.getByText(number)).toBeVisible()
  expect(screen.getByText(number)).toHaveAttribute('href', `/library/${number}`)
})

const manuscriptAttributes = [
  manuscript.provenance.parent ?? 'null',
  manuscript.provenance.name,
  manuscript.type.displayName ?? manuscript.type.name,
  manuscript.period.displayName ?? manuscript.period.name,
  manuscript.period.description,
]

test.each(manuscriptAttributes)('%s', async (attribute) => {
  userEvent.click(screen.getByText(manuscript.siglum))
  await waitFor(() => expect(screen.getByRole('tooltip')).toBeVisible())

  expect(screen.getByRole('tooltip')).toHaveTextContent(attribute)
})
