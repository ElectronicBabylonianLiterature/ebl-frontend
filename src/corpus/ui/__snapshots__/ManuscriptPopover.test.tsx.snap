// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show manuscript line details 1`] = `
<h3
  class="popover-header manuscript-popover__header"
>
  BabLBCompimofumew
  <span
    class="manuscript-popover__old-sigla"
  >
    Â (

    <span
      class="manuscript-popover__old-sigla__bibliography"
    >
      <span>
        fuwriv
      </span>
      <sup>

        <span
          class="reference-popover__citation"
        >
          Estrada & Wolfe, 2069: 7477711662481408-4847418310918144 
          [
          l. 4'.2., 3'.
          ]
          <span
            class="type-abbreviation"
          >
             (C)
          </span>
        </span>
      </sup>
    </span>
    )
  </span>
</h3>
`;
